src/calcium/qqbar.jl

```diff
@@ -0,0 +1,837 @@
+end
+
+function acospi(a::qqbar)
+   p = Vector{Int}(undef, 1)
+   q = Vector{Int}(undef, 1)
+   if !Bool(ccall((:qqbar_acos_pi, libcalcium), Cint, (Ptr{Int}, Ptr{Int}, Ref{qqbar}), p, q, a))
+      throw(DomainError(a, "function value is not algebraic"))
+   end
+   return qqbar(p[1]) // q[1]
+end
+
+function log_pi_i(a::qqbar)
+   p = Vector{Int}(undef, 1)
+   q = Vector{Int}(undef, 1)
+   if !Bool(ccall((:qqbar_log_pi_i, libcalcium), Cint, (Ptr{Int}, Ptr{Int}, Ref{qqbar}), p, q, a))
+      throw(DomainError(a, "function value is not algebraic"))
+   end
+   return qqbar(p[1]) // q[1]
+end
+
+
+
+###############################################################################
+#
+#   Guessing
+#
+###############################################################################
+
+function guess(R::CalciumQQBarField, x::acb, maxdeg::Int, maxbits::Int=0)
+   prec = precision(parent(x))
+   if maxbits <= 0
+      maxbits = prec
+   end
+   res = qqbar()
+   found = Bool(ccall((:qqbar_guess, libcalcium), Cint, (Ref{qqbar}, Ref{acb}, Int, Int, Int, Int), res, x, maxdeg, maxbits, 0, prec))
+   if !found
+      error("No suitable algebraic number found")
+   end
+   return res
+end
+
+function guess(R::CalciumQQBarField, x::arb, maxdeg::Int, maxbits::Int=0)
+   CC = AcbField(precision(parent(x)))
+   return guess(R, CC(x), maxdeg, maxbits)
+end
+
+###############################################################################
+#
+#   Conversions
+#
+###############################################################################
+
+function (R::ArbField)(a::qqbar)
+   prec = precision(R)
+   z = R()
+   ccall((:qqbar_get_arb, libcalcium), Nothing, (Ref{arb}, Ref{qqbar}, Int), z, a, prec)
+   !isfinite(z) && throw(DomainError(a, "nonreal algebraic number"))
+   return z
+end
+
+function (R::AcbField)(a::qqbar)
+   prec = precision(R)
+   z = R()
+   ccall((:qqbar_get_acb, libcalcium), Nothing, (Ref{acb}, Ref{qqbar}, Int), z, a, prec)
+   return z
+end
+
+# todo: provide qqbar_get_fmpq, qqbar_get_fmpz in C
+function fmpq(a::qqbar)
+   !isrational(a) && throw(DomainError(a), "nonrational algebraic number")
+   p = fmpz()
+   q = fmpz()
+   ccall((:fmpz_poly_get_coeff_fmpz, libflint), Nothing, (Ref{fmpz}, Ref{qqbar}, Int), p, a, 0)
+   ccall((:fmpz_poly_get_coeff_fmpz, libflint), Nothing, (Ref{fmpz}, Ref{qqbar}, Int), q, a, 1)
+   ccall((:fmpz_neg, libflint), Nothing, (Ref{fmpz}, Ref{fmpz}), p, p)
+   return p // q
+end
+
+function fmpz(a::qqbar)
+   !isinteger(a) && throw(DomainError(a), "noninteger algebraic number")
+   z = fmpz()
+   ccall((:fmpz_poly_get_coeff_fmpz, libflint), Nothing, (Ref{fmpz}, Ref{qqbar}, Int), z, a, 0)
+   ccall((:fmpz_neg, libflint), Nothing, (Ref{fmpz}, Ref{fmpz}), z, z)
+   return z
+end
+
+###############################################################################
+#
+#   Unsafe functions
+#
+###############################################################################
+
+function zero!(z::qqbar)
```

Comment:
We are starting to add tests for unsafe ops. See src/flint/fmpz_abs_series-test.jl for some example tests you could pinch.